observations/obs_converters/BATS/bats_to_obs.f90

```diff
@@ -0,0 +1,435 @@
+
+   read(input_line(date_firstcol:(date_firstcol + 3)), *, iostat=rcio) year
+   if(rcio /= 0) then
+      if(debug) print *, "got bad read code getting year at line ",line_number,", rcio = ",rcio
+      exit obsloop
+   end if
+
+   read(input_line((date_firstcol + 4):(date_firstcol + 5)), *, iostat=rcio) month
+   if(rcio /= 0) then
+      if(debug) print *, "got bad read code getting month at line ",line_number,", rcio = ",rcio
+      exit obsloop
+   end if
+
+   read(input_line((date_firstcol + 6):(date_firstcol + 7)), *, iostat=rcio) day
+   if(rcio /= 0) then
+      if(debug) print *, "got bad read code getting day at line ",line_number,", rcio = ",rcio
+      exit obsloop
+   end if
+
+   if((month == 02) .and. (day == 29)) then  ! we do not assimilate data taken on leap days
+      cycle obsloop
+   end if
+
+   ! extracting the time of day when the observation was taken
+
+   read(input_line(hourminute_firstcol:(hourminute_firstcol + 3)), *, iostat=rcio) hourminute_raw
+   if(rcio /= 0) then
+      if(debug) print *, "got bad read code parsing raw hour-minute at line ",line_number,", rcio = ",rcio
+      exit obsloop
+
+   else if(hourminute_raw == -999) then
+      cycle obsloop  ! missing timestamp
+
+   else if(hourminute_raw < 60) then
+      hour = 0
+
+   else
+      read(input_line(hourminute_firstcol:(hourminute_firstcol + 1)), *, iostat=rcio) hour
+
+      if(rcio /= 0) then
+         if(debug) print *, "got bad read code getting hour at line ",line_number,", rcio = ",rcio
+         exit obsloop
+
+      end if
+   end if
+
+   read(input_line((hourminute_firstcol + 2):(hourminute_firstcol + 3)), *, iostat=rcio) minute
+   if(rcio /= 0) then
+      if(debug) print *, "got bad read code getting minute at line ",line_number,", rcio = ",rcio
+      exit obsloop
+   end if
+
+   ! extracting the observation location
+
+   read(input_line(vert_cols(1):vert_cols(2)), *, iostat=rcio) vert
+   if(rcio /= 0) then
+      if(debug) print *, "got bad read code getting vert at line ",line_number,", rcio = ",rcio
+      exit obsloop
+
+   else if(abs(vert + 999.0) < 1.0d-8) then
+      cycle obsloop  ! missing vertical coordinate
+
+   else
+      vert = -vert   ! MOM6 depth coordinate becomes negative as you increase depth
+   end if
+
+   ! extracting the observation values
+
+   otype_loop : do otype_index = 1, NUM_SCALAR_OBS
+      read(input_line(scalar_obs_cols(1, otype_index):scalar_obs_cols(2, otype_index)), *, iostat=rcio) ovalue
+      
+      if(rcio /= 0) then
+         if(debug) print *, "got bad read code getting observation type ",otype_index," at line ",line_number,", rcio = ",rcio
+         exit obsloop
+
+      else if(abs(ovalue + 999.0) < 1.0d-8) then
+         cycle otype_loop ! missing observation
+
+      end if
+
+      ! unit corrections from BATS to MARBL
+      if((OTYPE_ORDERING(otype_index) == BATS_ALKALINITY) .or. (OTYPE_ORDERING(otype_index) == BATS_INORGANIC_CARBON)) then
+         ovalue = ovalue*1.026
```

Comment:
just a comment, 1.026 is a wild unit conversion.

Reply:
agreed :)